### PR TITLE
Message TTL

### PR DESCRIPTION
**Pull request by k-wall**
### Type of change

- Enhancement / new feature

### Description

The first part of this change changes the standard controller / agent (brokered path) to maintain the address.status.ttl portion of the address object with the resulting TTL clamp defined by the address plan or address itself.

The second part, which is common to the standard and brokered paths, propagates the desired TTL settings (from address.status.ttl) to  the address settings.

### Checklist

- [ ] Update/write design documentation in `./documentation/design`
- [x] Write tests and make sure they pass
- [ ] Update documentation
- [x] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [x] Update CHANGELOG.md

### CODE REVIEW

**Comment by k-wall** [comment]
@enmasse-ci run tests profile=isolated

File: agent/lib/internal_address_source.js [hunk 1]
```diff
@@ -98,9 +98,15 @@ function same_address_definition(a, b) {
     return a.address === b.address
         && a.type === b.type
         && same_allocation(a.allocated_to, b.allocated_to)
+        && same_ttl(a.ttl, b.ttl)
         && same_plan_status(a.status ? a.status.planStatus : undefined, b.status ? b.status.planStatus : undefined);
 }
 
+function same_ttl(a, b) {
+    if (a === undefined) return b === undefined;
+    return b && a.minimum === b.minimum && a.maximum === b.maximum;
```
Comment (lulf) [inline agent/lib/internal_address_source.js:107]: Even though a !== undefined, could a still be null?

File: agent/lib/internal_addressplan_source.js [hunk 2]
```diff
@@ -89,9 +89,7 @@ function same_addressplan_resources(a, b) {
 
 function same_ttl(a, b) {
```
Comment (lulf) [inline agent/lib/internal_addressplan_source.js:90]: Maybe reuse function from internal_address_source

File: broker-plugin/plugin/src/main/sh/init-broker.sh [hunk 3]
```diff
@@ -8,6 +8,8 @@ BROKER_PLUGIN_DIR=/opt/broker-plugin
 BROKER_CUSTOM=${BROKER_DIR}/custom
 CONFIG_TEMPLATES=${BROKER_PLUGIN_DIR}/conf
 BROKER_CONF_DIR=${BROKER_CUSTOM}/conf
+MESSAGE_EXPIRY_SCAN_PERIOD=${MESSAGE_EXPIRY_SCAN_PERIOD:-30000}
```
Comment (lulf) [inline broker-plugin/plugin/src/main/sh/init-broker.sh:11]: Should we expose it in infra config?

File: agent/lib/broker_controller.js [hunk 4]
```diff
@@ -432,22 +432,29 @@ BrokerController.prototype.update_address_setting = function (a) {
...
                 } else {
                     log.debug('[%s] Address settings match for %s: not updating', self.id, name);
                     return Promise.resolve();
                 }
+            } else if (compare_address_settings(orig_settings, self.root_address_settings) !== 0) {
+                // Settings were previously applied but now no longer required.  To remove them first we
+                // reset the address settings to the root values then remove the record.
+                log.info('[%s] Removing address settings: %s', self.id, name);
+                return self.broker.addAddressSettings(name, self.root_address_settings).finally(() => {
+                    return self.broker.removeAddressSettings(name);
```
Comment (lulf) [inline agent/lib/broker_controller.js:446]: Yikes :) We should try to patch the broker to make this simpler to handle in 1.0 :)

Comment (k-wall) [reply agent/lib/broker_controller.js:446]: yeah - there's an oddity with Artemis's behaviour, I'll raise an issue.

File: systemtests/src/test/java/io/enmasse/systemtest/isolated/ttl/TtlTest.java [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+import static io.enmasse.systemtest.TestTag.ISOLATED;
+import static org.hamcrest.CoreMatchers.*;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
+import static org.junit.jupiter.api.Assertions.*;
+
+@Tag(ISOLATED)
```
Comment (kornys) [inline systemtests/src/test/java/io/enmasse/systemtest/isolated/ttl/TtlTest.java:47]: Should be added also profile acceptance if we want to see behavior on ocp4

File: broker-plugin/plugin/src/main/sh/init-broker.sh (hunk 3 above)
Comment (k-wall) [reply broker-plugin/plugin/src/main/sh/init-broker.sh:11]: @lulf I'm not sure, I don't think most customers would want to tune this and I would actively discourage business use-cases that relied on timely TTL (that's not its intended purpose).  I'd be happy to add it.

Comment (lulf) [reply broker-plugin/plugin/src/main/sh/init-broker.sh:11]: I think its ok to leave it out for now.

File: systemtests/src/test/java/io/enmasse/systemtest/isolated/ttl/TtlTest.java [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+
+        awaitAddressSettingsSync(addressSpace, addrWithTtl, new Ttl());
+
+        try(AmqpClient client = getAmqpClientFactory().createQueueClient(addressSpace)) {
+            client.getConnectOptions().setCredentials(user);
+
+            List<Message> messages = new ArrayList<>();
+            AtomicInteger count = new AtomicInteger();
+            count.set(0);
+            IntStream.of(1).forEach(unused -> {
```
Comment (kornys) [inline systemtests/src/test/java/io/enmasse/systemtest/isolated/ttl/TtlTest.java:247]: is it necessary to use foreach with one loop?

File: agent/lib/internal_addressplan_source.js (hunk 2 above)
Comment (k-wall) [reply agent/lib/internal_addressplan_source.js]: Hopefully I've addressed this criticism in the latest code.

File: agent/lib/internal_address_source.js (hunk 1 above)
Comment (k-wall) [reply agent/lib/internal_address_source.js]: Should be fixed.

**Comment by k-wall** [comment]
@enmasse-ci run tests profile=isolated

**Comment by k-wall** [comment]
@enmasse-ci run tests profile=isolated

**Comment by k-wall** [comment]
@enmasse-ci run tests profile=isolated

**Comment by k-wall** [comment]
@enmasse-ci run tests ocp4 profile=isolated

**Comment by k-wall** [comment]
@enmasse-ci run tests ocp4

File: templates/crds/addressplans.crd.yaml [hunk 7]
```diff
@@ -90,6 +90,16 @@ spec:
           type: string
         addressType:
           type: string
+        ttl:
+          type: object
+          description: "Restricts message time-to-live (TTL). Applies to address types queue and topic only."
```
Comment (lulf) [inline templates/crds/addressplans.crd.yaml:95]: The description should include the expected unit for maximum and minimum values.

Comment (k-wall) [reply templates/crds/addressplans.crd.yaml:95]: Done.

**Comment by k-wall** [comment]
@enmasse-ci run tests profile=isolated testcase=MessageTtlTest

**Pull request merged**